src/osp/Resource/Resource.h

```diff
@@ -122,19 +122,15 @@ class DependRes
 
     DependRes<TYPE_T>& operator=(DependRes<TYPE_T>&& move)
     {
-        // TODO: this sucks, please change it
-        if (!move.m_empty)
-        {
-            move.m_it->second.m_refCount --;
-        }
         if (!m_empty)
         {
-            // existing iterator is being overwritten, release it
-            m_it->second.m_refCount --;
+            // Existing iterator is being overwritten, decrease ref count
+            m_it->second.m_refCount--;
         }
-        m_empty = false;
+        m_empty = move.m_empty;
+        move.m_empty = true;
```

Comment:
instead of m_empty, maybe there's room for std::optional<> to be used in part of this?

Reply:
Definitely. m_empty just tracks if its internal std::map iterator is valid or not. I guess using std::optional changes it to "valid iterator or nothing."